uwrt_arm_cartesian_controller/src/uwrt_arm_cartesian_controller.cpp

```diff
@@ -0,0 +1,224 @@
+#include <array>
+
+namespace uwrt_arm_cartesian_controller
+{
+    UWRTCartesianController::UWRTCartesianController() : m_cart_command_sub(nullptr), m_ik_solver(nullptr) {}
+
+    CallbackReturn UWRTCartesianController::on_init()
+    {
+        try
+        {
+            auto_declare<std::vector<std::string>>("joint_names", std::vector<std::string>());
+            auto_declare<std::string>("interface_name", "");
+            auto_declare<std::string>("end_effector_velocity_topic", "");
+            auto_declare<std::string>("robot_description", "");
+            auto_declare<std::string>("root_name", "");
+            auto_declare<std::string>("tip_name", "");
+        }
+        catch (const std::exception &e)
+        {
+            fprintf(stderr, "Exception thrown during init stage with message: %s \n", e.what());
+            return CallbackReturn::ERROR;
+        }
+
+        return CallbackReturn::SUCCESS;
+    }
+
+    CallbackReturn UWRTCartesianController::on_configure(
+        const rclcpp_lifecycle::State & /*previous_state*/)
+    {
+        if (!parse_params())
+        {
+            return CallbackReturn::ERROR;
+        }
+
+        // Generate tree from urdf
+        if (!kdl_parser::treeFromParam(m_robot_description, m_kdl_tree))
+        {
+            RCLCPP_ERROR(get_node()->get_logger(), "Failed to parse KDL tree from robot description");
+            return CallbackReturn::ERROR;
+        }
+
+        // Populate the KDL chain
+        const bool chain_generated_successfully = m_kdl_tree.getChain(m_root_name, m_tip_name, m_kdl_chain);
+        std::stringstream tree_info = "Failed to get KDL chain from tree: \n\t" << m_root_name << " --> " << m_tip_name;
+        tree_info << "\n\tTree has " << m_kdl_tree.getNrOfJoints() << " joints \n\tThe joint names are:";
+        for (auto it = m_kdl_chain.segments.begin(); it != m_kdl_chain.segments.end(); ++it)
+        {
+            tree_info << "\n\t\t" << it->getJoint().getName());
+        }
+        tree_info << "\n\tTree has " << m_kdl_tree.getNrOfSegments() << " segments"
+                  << "\n\tThe segments are:";
+        KDL::SegmentMap segment_map = m_kdl_tree.getSegments();
+        for (auto it = segment_map.begin(); it != segment_map.end(); it++)
+        {
+            tree_info << "\n\t\t" << it->first);
+        }
+
+        if (!chain_generated_successfully)
+        {
+            RCLCPP_ERROR(get_node()->get_logger(), tree_info.str());
+            return CallbackReturn::ERROR;
+        }
+
+        RCLCPP_INFO(get_node()->get_logger(), tree_info.str());
+
+        ROS_INFO_STREAM("***Listing HW Interfaces: ***");
+        // Get joint handles for all of the joints in the chain
+        /*for (std::vector<KDL::Segment>::const_iterator it = kdl_chain_.segments.begin(); it != kdl_chain_.segments.end(); ++it)
+        { younes todo wut is this? wut is joing handles wut is command_interfaces
+            // WE DON'T WANT AN INTERFACE FOR THE FIXED JOINT
```

Comment:
this is some cruft that was specific to the old arm's urdf. it had this fixed joint(arm_shoulder_bicep_joint) because the arm had a fixed joint(the weird L-shaped piece) that was near the linear actuator.